ext/mbstring/libmbfl/filters/mbfilter_uuencode.c

```diff
@@ -250,49 +250,121 @@ static void mb_wchar_to_uuencode(uint32_t *in, size_t len, mb_convert_buf *buf,
 	/* Every 3 bytes of input gets encoded as 4 bytes of output
 	 * Additionally, we have a 'length' byte and a newline for each line of output
 	 * (Maximum 45 input bytes can be encoded on a single output line)
-	 * Make space for two more bytes in case we start close to where a line must end */
-	MB_CONVERT_BUF_ENSURE(buf, out, limit, ((len + 2) * 4 / 3) + (((len + 44) / 45) * 2) + (buf->state ? 0 : sizeof("begin 0644 filename\n")) + 2);
+	 * Make space for two more bytes in case we start close to where a line must end,
+	 * and another two if there are cached bits remaining from the previous call */
+	MB_CONVERT_BUF_ENSURE(buf, out, limit, ((len + 2) * 4 / 3) + (((len + 44) / 45) * 2) + (buf->state ? 0 : sizeof("begin 0644 filename\n")) + 4);
 
-	unsigned int bytes_encoded = buf->state >> 1;
+	unsigned int bytes_encoded = (buf->state >> 1) & 0x7F;
+	/* UUEncode naturally wants to process input bytes in groups of 3, but again,
+	 * our buffer size may not be a multiple of 3
+	 * So there may be data from the previous call which we need to flush out */
+	unsigned int n_cached_bits = (buf->state >> 8) & 0xFF;
+	unsigned int cached_bits = buf->state >> 16;
 
 	if (!buf->state) {
 		for (char *s = "begin 0644 filename\n"; *s; s++) {
 			out = mb_convert_buf_add(out, *s);
 		}
 		out = mb_convert_buf_add(out, MIN(len, 45) + 32);
 		buf->state |= 1;
+	} else if (!len && end && !bytes_encoded && !n_cached_bits) {
+		/* Corner case: under EXTREMELY rare circumstances, it's possible that the
+		 * final call to this conversion function will happen with an empty input
+		 * buffer, leaving an unwanted trailing len byte in the output buffer.
```

Comment:
It requires two different events to coincide.

First, please note that whenever we emit a UUEncode line ending, we eagerly emit the length byte for the next line, as long as it *seems* that more input is still coming (i.e. `end` is false). So the first event is that we eagerly emit a UUEncode line ending and len byte, and then immediately return (i.e. a buffer of wchars is finished).

Then the second event is that the following buffer of wchars is the last one, and *contains zero wchars*. Why would this happen? Well, remember that the `to_wchar` filters generally iterate over the input string and emit wchars *until the wchar buffer is full*. If there are still more input bytes to process, then we will re-enter the `to_wchar` function later. However, when we re-enter the `to_wchar` function, we may discover that there are actually no wchars remaining. This can happen if the few bytes remaining are something like a JIS escape sequence, which shifts the JIS decoder into a different mode but doesn't actually decode to anything.

If both of those events unluckily coincide, then our UUEncode encoder could leave an extraneous trailing byte in its output buffer... if it were not for this `if` clause.